w3af/plugins/grep/vulners_scanner.py

```diff
@@ -0,0 +1,219 @@
+            self._multi_re = MultiRE(((regex, regexAliases.get(regex)) for regex in regexAliases), re.IGNORECASE)
+        except Exception as e:
+            self.rules_table = None
+            error_message = 'Vulners plugin failed to init with error: %s'
+            om.out.error(error_message % e)
+
+    def get_vulners_api(self):
+        # Lazy import. Just not to make it in the __init__
+        # Other way API will try to communicate Vulners right at the w3af startup
+        if not self._vulners_api:
+            # Don't forget to setup Vulners API key if you have one
+            self._vulners_api = vulners.Vulners(api_key= self._vulners_api_key)
+        return self._vulners_api
+
+    def check_vulners(self, software_name, software_version, check_type):
+        vulnerabilities = {}
+        cached_result = self._vulnerability_cache.get((software_name, software_version, check_type))
+        if cached_result:
+            return cached_result
+        # Ask Vulners about vulnerabilities
+        if check_type == 'software':
+            vulnerabilities = self.get_vulners_api().softwareVulnerabilities(software_name, software_version)
+        elif check_type == 'cpe':
+            cpe_string = "%s:%s" % (software_name, software_version)
+            vulnerabilities = self.get_vulners_api().cpeVulnerabilities(cpe_string.encode())
+        self._vulnerability_cache[(software_name, software_version, check_type)] = vulnerabilities
+        return vulnerabilities
+
+    def get_severity(self, cvss_score):
+        # This table can be actually modified
+        # That's only my opinions on the severity vs CVSS.score
+        # Maybe move it to the options later?
+        cvss_score = cvss_score * 10
+        if cvss_score in range(20, 30):
+            return severity.LOW
+        elif cvss_score in range(30, 70):
+            return severity.MEDIUM
+        elif cvss_score in range(70, 100):
+            return severity.HIGH
+        else:
+            return severity.INFORMATION
+
+    def grep(self, request, response):
+        """
+        Plugin entry point, search for vulnerable software banners  in web application HTTP responses.
+        :param request: The HTTP request object.
+        :param response: The HTTP response object
+        :return: None
+
+        """
+        # Lazy update rules if it's first start of the plugin
+        if not self.rules_updated:
+            self.update_vulners_rules()
+            self.rules_updated = True
+
+        # Check if we have downloaded rules well.
+        # If there is no rules - something went wrong, time to exit.
+        if not self.rules_table:
+            return
+        # We do not parse non-text output
+        if not response.is_text_or_html():
+            return
+        if response.get_url().get_domain_path() in self._already_visited:
+            return
+
+        self._already_visited.add(response.get_url().get_domain_path())
+
+        raw_response = response.dump()
```

Comment:
Do you guys need to match the whole HTTP response? Including the body? What if we just match the regex against headers?

My fear is that this will at some point leak sensitive information to vulners.com. How?
 * A broken regex is obtained by w3af in `update_vulners_rules`. 
 * This regex matches more than required and matches sensitive information (cookie, response body, etc.)
 * w3af plugin sends that to vulners.com in `check_vulners` `software_version` parameter

I would feel much better if we could just send the HTTP response headers to `self._multi_re.query` below, and before doing that we clean the sensitive headers like set-cookie.

Reply:
Wow! Fantastic scenario, never thought about it.
We are using this rules: https://github.com/vulnersCom/detect-rules
They also matches http body and JS scripts and so on. Not only headers.
I think the best way will be:
1) You will fork them to w3af github account.
2) Plugin will take them from yours github.
3) We will make time-to-time update of them, and you will pull updates validating them.

I've understand that it will take a little time to maintain, but users will bu sure what kind of regexes are used. Privacy first, yep.

Reply:
Sadly, and because of licensing issues, I can't include rules.json file in the w3af repository.

rules.json is GPLv3
w3af is GPLv2

Those licenses are incompatible. I can't distribute GPLv3 licensed files inside my GPLv2 project.

A potential solution would be for w3af to download https://raw.githubusercontent.com/vulnersCom/detect-rules/master/rules.json at the beginning of the scan. You could do it similar to https://github.com/andresriancho/w3af/blob/develop/w3af/plugins/grep/retirejs.py#L148-L167

I trust https://raw.githubusercontent.com/vulnersCom/detect-rules/master/rules.json more than the result of the vulners.com API because:
 * There is a git log which we could potentially use to find someone modifying the JSON
 * Users can audit it if they want to, knowing that whatever is in that URL can not be changed unless the git log changes

Reply:
Ok! Will do it that way.
But looks like I need to checkout at develop branch, because it master grep plugin does not have "self._uri_opener".